File: stroke_extractor.py
'''
Given an svg.path.Path object representing a glyph, a StrokeExtractor instance
will break it down into a list of svg.path.Path objects, one for each stroke.

The algorithm we currently use is a 'corner-and-bridge' algorithm. First, we
detect possible corners in the path object. 'Corners' are points where the
derivative of the curve angle is sharply negative - that is, points at which
the curve is very non-convex. If two strokes cross eachother, we should detect
four corners, one at each place at the outline of the intersection.

(Note that much more complex configurations are possible - for example a stroke
may end at the middle of another stroke, or many strokes may intersect to form
a star shape.)

We then detect 'bridges', which are edges between corners where the stroke
entering one corner may continue to the stroke exiting the other corner. In our
two-strokes-crossing example, we should detect four bridges connecting the four
corners to form a simple quadrilateral.

Finally, we traverse the path, usually following SVG path elements, but taking
bridges when they are inline with the previously traversed path element. The
output of this traversal is our final stroke decomposition.

At many points during this algorithm we may detect various anomalies. We log
these anomalies so that they can be reviewed manually.
'''
import collections
import math
import svg.path


MAX_BRIDGE_DISTANCE = 128
MAX_CORNER_MERGE_DISTANCE = 16
MIN_CORNER_ANGLE = 0.1*math.pi
MIN_CORNER_TANGENT_DISTANCE = 4


def area(path):
  '''
  Returns the area of the path. The result is positive iff the path winds in
  the counter-clockwise direction.
  '''
  def area_under_curve(x):
    return (x.start.real - x.end.real)*(x.start.imag + x.end.imag)
  return int(sum(map(area_under_curve, path))/2)


def split_and_orient_path(path):
  '''
  Takes a non-empty svg.path.Path object that may contain multiple closed loops.
  Returns a list of svg.path.Path objects that are all minimal closed curve.
  The returned paths will be the way a TTF glyph should be: exterior curves
  will be counter-clockwise and interior curves will be clockwise.
  '''
  paths = [[path[0]]]
  for element in path[1:]:
    if element.start == element.end:
      continue
    if element.start != paths[-1][-1].end:
      paths.append([])
    paths[-1].append(element)
  # Determine if this glyph is oriented in the wrong direction by computing the
  # area of each glyph. The glyph with maximum |area| should have positive area,
  # because it must be an exterior path.
  def reverse(path):
    for element in path:
      (element.start, element.end) = (element.end, element.start)
    return reversed(path)
  areas = [area(path) for path in paths]
  max_area = max((abs(area), area) for area in areas)[1]
  if max_area < 0:
    paths = map(reverse, paths)
  return [svg.path.Path(*path) for path in paths]


class Corner(object):
  def __init__(self, paths, index):
    self.index = index
    (i, j) = index
    self.path = paths[i]
    self.point = paths[i][j].start
    (self.tangent1, self.tangent2) = self._get_tangents()
    self.angle = self._get_angle(self.tangent1, self.tangent2)

  def bridge(self, other):
    '''
    Returns true if a stroke continues from this corner point to the other.
    Internally, this function builds a 7-dimensional feature vector and then
    calls a classifier. The 7 features are:
      features[0]: The angle between the edge in and the bridge
      features[1]: The angle between the bridge and the edge out
      features[2]: The angle between the cross stroke out and the bridge
      features[3]: The angle between the cross stroke in and the bridge
      features[4]: The angle at this corner
      features[5]: The angle at the other corner
      features[6]: The length of the bridge

    At an ideal bridge, features[0] and features[1] should be very close to 0,
    meaning that the stroke can continue smoothly from this corner to the other.
    features[2] + features[3] is close to pi, meaning that the stroke in
    is straight, and features[6], the distance, is small.

    This ideal configuration might look like this diagram:

            /  ^
           /  /
        <-O  S--

    where S is this corner and O is the other and the arrows indicate the
    direction of the curve.
    '''
    diff = other.point - self.point
    length = abs(diff)
    if length == 0 or length > MAX_BRIDGE_DISTANCE:
      return False
    # NOTE: These angle features make sense even if points are on different
    # subpaths of the glyph path! Because of our preprocessing, exterior glyph
    # paths are clockwise while interior paths are counter-clockwise, so angle
    # features around a bridge are the same whether or not the two sides of
    # the bridge are on the same path.
    features = (
      self._get_angle(self.tangent1, diff),
      self._get_angle(diff, other.tangent2),
      self._get_angle(diff, self.tangent2),
      self._get_angle(other.tangent1, diff),
      self.angle,
      other.angle,
      length,
    )
    # TODO(skishore): Log this sample and use it to train the classifier.
    result = self._run_classifier(features)
    return result

  def merge_into(self, other):
    '''
    Merges this corner into the other corner, updating the other's data.
    The merged corner takes the position of the sharper corner of the two.
    Because the path curves slightly in the positive direction on average, a
    curve is sharper if its angle is more negative.
    '''
    if self.angle < other.angle:
      other.index = self.index
      other.point = self.point
    other.tangent1 = self.tangent1
    other.angle = other._get_angle(other.tangent1, other.tangent2)

  def should_merge(self, other):
    '''
    Returns true if this corner point is close enough to the next one that
    they should be combined into one corner point. Note that the next corner
    should have an index that occurs soon after this corner's.
    '''
    assert other.index[0] == self.index[0], \
           'merge called for corners on different curves!'
    if abs(other.point - self.point) > MAX_CORNER_MERGE_DISTANCE:
      return False
    distance = 0
    j = self.index[1]
    while j != other.index[1]:
      distance += abs(self.path[j].end - self.path[j].start)
      j = (j + 1) % len(self.path)
    return distance < MAX_CORNER_MERGE_DISTANCE

  def _get_angle(self, vector1, vector2):
    ratio = vector2/vector1 if vector1 else 0
    return math.atan2(ratio.imag, ratio.real)

  def _get_tangents(self):
    segment1 = self.path[self.index[1] - 1]
    tangent1 = segment1.end - segment1.start
    if (type(segment1) == svg.path.QuadraticBezier and
        abs(segment1.end - segment1.control) > MIN_CORNER_TANGENT_DISTANCE):
      tangent1 = segment1.end - segment1.control
    segment2 = self.path[self.index[1]]
    tangent2 = segment2.end - segment2.start
    if (type(segment2) == svg.path.QuadraticBezier and
        abs(segment2.control - segment2.start) > MIN_CORNER_TANGENT_DISTANCE):
      tangent2 = segment2.control - segment2.start
    return (tangent1, tangent2)

  def _run_classifier(self, features):
    # TODO(skishore): Replace these inequalities with a trained classifier.
    alignment = abs(features[0]) + abs(features[1])
    incidence = abs(features[2] + features[3] + math.pi)
    short = features[6] < MAX_BRIDGE_DISTANCE/2
    clean = alignment < 0.1*math.pi or alignment + incidence < 0.2*math.pi
    cross = all([
      features[0] > 0,
      features[1] > 0,
      features[2] + features[3] < -0.5*math.pi,
    ])
    result = 0
    if features[2] < 0 and features[3] < 0 and (clean or (short and cross)):
      result = (1 if short else 0.75) if clean else 0.5
    return result


class StrokeExtractor(object):
  def __init__(self, name, d):
    self.name = name
    self.messages = []
    self.paths = split_and_orient_path(svg.path.parse_path(d))
    self.corners = self.get_corners()
    self.bridges = self.get_bridges()
    (self.strokes, self.stroke_adjacency) = self.extract_strokes()

  def extract_stroke(self, extracted, start):
    '''
    Given a path, a list of corners, and an adjacency list representation of
    bridges between then, extract a stroke that starts at the given index
    and add the indices of all elements on that stroke to extracted.

    This method will return a pair (path, corners), where the first element is
    an svg.path.Path object representing the stroke and the second is a list of
    corners that appear on that stroke. The corners list will have duplicates if
    the stroke loops back on itself, which indicates a mistake somewhere.

    This method will fail if, when following edges the the initial path element,
    we cross a bridge and enter a stroke that has already been extracted. If so,
    the path we return will be None.

    NOTE: We deliberately avoid using bridge directionality in this algorithm
    so that we can handle manually added bridges.
    '''
    current = start
    corners = []
    path = svg.path.Path()
    visited = set()

    def advance(index):
      return (index[0], (index[1] + 1) % len(self.paths[index[0]]))

    def angle(index, bridge):
      tangent = self.corners[index].tangent1
      ratio = (self.corners[bridge].point - self.corners[index].point)/tangent
      return abs(math.atan2(ratio.imag, ratio.real))

    while True:
      # Add the current stroke element to the path and advance along it.
      path.append(self.paths[current[0]][current[1]])
      visited.add(current)
      current = advance(current)
      # If there is a bridge aligned with the stroke element that we advanced
      # over, advance over that bridge as well. If there are multiple bridges,
      # choose the one that is most aligned.
      if current in self.bridges:
        next = sorted(self.bridges[current], key=lambda x: angle(current, x))[0]
        corners.extend([self.corners[current], self.corners[next]])
        path.append(svg.path.Line(
            start=self.corners[current].point, end=self.corners[next].point))
        current = next
      # Check if we either closed the loop or hit an already extracted stroke.
      if current == start:
        extracted.update(visited)
        return (path, corners)
      elif current in visited or current in extracted:
        return (None, [])

  def extract_strokes(self):
    '''
    Returns a pair (strokes, stroke_adjacency), where the first element is a
    list of svg.path.Path objects that decompose this glyph into strokes and the
    second is an adjacency-list representation of the indices of strokes which
    share corner points.

    This method will log if some path elements do not appear on any stroke.
    '''
    extracted = set()
    strokes = []
    stroke_adjacency = collections.defaultdict(set)
    corner_adjacency = collections.defaultdict(set)
    for i, path in enumerate(self.paths):
      for j, element in enumerate(path):
        index = (i, j)
        if index not in extracted:
          (stroke, corners) = self.extract_stroke(extracted, index)
          if stroke is None:
            self.log('Stroke extraction missed some path elements!')
            continue
          stroke_index = len(strokes)
          strokes.append(stroke)
          corner_indices = set(corner.index for corner in corners)
          if len(corner_indices) < len(corners):
            self.log('Stroke {0} is self-intersecting!'.format(stroke_index))
          for corner_index in corner_indices:
            for other_index in corner_adjacency[corner_index]:
              stroke_adjacency[other_index].add(stroke_index)
              stroke_adjacency[stroke_index].add(other_index)
            corner_adjacency[corner_index].add(stroke_index)
    return (strokes, stroke_adjacency)

  def get_bridges(self):
    '''
    Returns an adjacency list of bridges. A bridge is a pair of corner indices
    through which a stroke continues. The adjacency list is undirected: for any
    two corner indices a and b, if b in result[a], a in result[b].
    '''
    # Collect bridge candidates scored by our bridge classifier.
    candidates = []
    for corner in self.corners.itervalues():
      for other in self.corners.itervalues():
        confidence = corner.bridge(other)
        if confidence > 0:
          candidates.append((confidence, corner.index, other.index))
    candidates.sort(reverse=True)
    # Add bridges to the set of bridges in order of decreasing confidence.
    # However, we do NOT add bridges that would either a) form a triangle with
    # an existing bridge, or b) that are long and should be multiple bridges.
    bridges = set()
    for (confidence, index1, index2) in candidates:
      other1 = set(b for (a, b) in bridges if a == index1)
      other2 = set(b for (a, b) in bridges if a == index2)
      if (other1.intersection(other2) or
          self.should_split_bridge((index1, index2))):
        continue
      bridges.add((index1, index2))
      bridges.add((index2, index1))
    # Convert the result to an adjacency list. Having more than two bridges at
    # any given corner results in a warning.
    result = collections.defaultdict(list)
    for (index1, index2) in bridges:
      result[index1].append(index2)
      if len(result[index1]) == 3:
        self.log('More than two bridges at corner {0}'.format(
            self.corners[index1].point))
    return result

  def get_corners(self):
    '''
    Returns a dict mapping indices to corners at that index. Each corner is a
    point on the curve where the path makes a sharp negative angle. Since the
    path has a small positive average angle, it is non-convex at these corners.
    '''
    result = {}
    for i, path in enumerate(self.paths):
      candidates = [Corner(self.paths, (i, j)) for j in xrange(len(path))]
      j = 0
      while j < len(candidates):
        next_j = (j + 1) % len(candidates)
        if candidates[j].should_merge(candidates[next_j]):
          candidates[j].merge_into(candidates[next_j])
          candidates.pop(j)
        else:
          j += 1
      for corner in filter(lambda x: x.angle < -MIN_CORNER_ANGLE, candidates):
        result[corner.index] = corner
    return result

  def log(self, message):
    self.messages.append(message)

  def should_split_bridge(self, bridge):
    '''
    Returns true if there is some corner that is too close to the middle of the
    given bridge. When this occurs, the gap between these indices should usually
    be spanned by multiple bridges instead.
    '''
    (index1, index2) = bridge
    base = self.corners[index1].point
    diff = self.corners[index2].point - base
    for corner in self.corners.itervalues():
      if corner.index in bridge:
        continue
      t = ((corner.point.real - base.real)*diff.real +
           (corner.point.imag - base.imag)*diff.imag)/(abs(diff)**2)
      distance_to_line = abs(self.corners[index1].point + t*diff - corner.point)
      if 0 < t < 1 and distance_to_line < MAX_CORNER_MERGE_DISTANCE:
        return True
    return False

File: lib/stroke_extractor.js
import {assert, Angle, Point} from '/lib/base';
import {Hungarian} from '/lib/hungarian';
import {svg} from '/lib/svg';

const MAX_BRIDGE_DISTANCE = 64;
const MIN_CORNER_ANGLE = 0.1*Math.PI;
const MIN_CORNER_TANGENT_DISTANCE = 4;
const REVERSAL_PENALTY = 0.5;

// Errors out if the bridges are invalid in some gross way.
const checkBridge = (bridge) => {
  assert(Point.valid(bridge[0]) && Point.valid(bridge[1]));
  assert(!Point.equal(bridge[0], bridge[1]));
}

// Returns the list of bridges on the path with the given endpoints. We strip
// nearly all of the metadata out of this list to make it easy to hand-correct.
// The list that we return is simply a list of pairs of points.
const getBridges = (endpoints, classifier) => {
  const result = [];
  const corners = endpoints.filter((x) => x.corner);
  const matching = matchCorners(corners, classifier);
  for (let i = 0; i < corners.length; i++) {
    const j = matching[i];
    if (j <= i && matching[j] === i) {
      continue;
    }
    result.push([Point.clone(corners[i].point), Point.clone(corners[j].point)]);
  }
  result.map(checkBridge);
  return result;
}

// Returns a list of angle and distance features between two corners.
const getFeatures = (ins, out) => {
  const diff = Point.subtract(out.point, ins.point);
  const trivial = Point.equal(diff, [0, 0]);
  const angle = Math.atan2(diff[1], diff[0]);
  const distance = Math.sqrt(Point.distance2(out.point, ins.point));
  return [
    Angle.subtract(angle, ins.angles[0]),
    Angle.subtract(out.angles[1], angle),
    Angle.subtract(ins.angles[1], angle),
    Angle.subtract(angle, out.angles[0]),
    Angle.subtract(ins.angles[1], ins.angles[0]),
    Angle.subtract(out.angles[1], out.angles[0]),
    (trivial ? 1 : 0),
    distance/MAX_BRIDGE_DISTANCE,
  ];
}

// A hand-tuned classifier that uses the features above to return a score for
// connecting two corners by a bridge. This classifier throws out most data.
const handTunedClassifier = (features) => {
  if (features[6] > 0) {
    return -Angle.penalty(features[4]);
  }
  let angle_penalty = Angle.penalty(features[0]) + Angle.penalty(features[1]);
  const distance_penalty = features[7];
  if (features[0] > 0 && features[1] > 0 &&
      features[2] + features[3] < -0.5*Math.PI) {
    angle_penalty = angle_penalty/16;
  }
  return -(angle_penalty + distance_penalty);
}

// Takes a list of corners and returns a bipartite matching between them.
// If matching[i] === j, then corners[i] is matched with corners[j] - that is,
// we should construct a bridge from corners[i].point to corners[j].point.
const matchCorners = (corners, classifier) => {
  const matrix = [];
  for (let i = 0; i < corners.length; i++) {
    matrix.push([]);
    for (let j = 0; j < corners.length; j++) {
      matrix[i].push(scoreCorners(corners[i], corners[j], classifier));
    }
  }
  for (let i = 0; i < corners.length; i++) {
    for (let j = 0; j < corners.length; j++) {
      const reversed_score = matrix[j][i] - REVERSAL_PENALTY;
      if (reversed_score > matrix[i][j]) {
        matrix[i][j] = reversed_score;
      }
    }
  }
  return (new Hungarian(matrix)).x_match;
}

// Takes two corners and returns the score assigned to constructing a bridge
// from one corner to the other. The score is directed: the bridge from ins to
// out may be weighted higher than from out to ins.
const scoreCorners = (ins, out, classifier) => {
  return classifier(getFeatures(ins, out));
}

// Stores angle and distance metadata around an SVG path segment's start point.
// This endpoint may be a 'corner', which is true if the path bends sharply in
// the negative (clockwise) direction at that point.
function Endpoint(paths, index) {
  this.index = index;
  const path = paths[index[0]];
  const n = path.length;
  this.indices = [[index[0], (index[1] + n - 1) % n], index];
  this.segments = [path[(index[1] + n - 1) % n], path[index[1]]];
  this.point = this.segments[0].end;
  assert(Point.valid(this.point), this.point);
  assert(Point.equal(this.point, this.segments[1].start), path);
  this.tangents = [
    Point.subtract(this.segments[0].end, this.segments[0].start),
    Point.subtract(this.segments[1].end, this.segments[1].start),
  ];
  const threshold = Math.pow(MIN_CORNER_TANGENT_DISTANCE, 2);
  if (this.segments[0].control !== undefined &&
      Point.distance2(this.point, this.segments[0].control) > threshold) {
    this.tangents[0] = Point.subtract(this.point, this.segments[0].control);
  }
  if (this.segments[1].control !== undefined &&
      Point.distance2(this.point, this.segments[1].control) > threshold) {
    this.tangents[1] = Point.subtract(this.segments[1].control, this.point);
  }
  this.angles = this.tangents.map(Point.angle);
  const diff = Angle.subtract(this.angles[1], this.angles[0]);
  this.corner = diff < -MIN_CORNER_ANGLE;
  return this;
}

// Code for the stroke extraction step follows.

const addEdgeToAdjacency = (edge, adjacency) => {
  assert(edge.length === 2);
  adjacency[edge[0]] = adjacency[edge[0]] || [];
  if (adjacency[edge[0]].indexOf(edge[1]) < 0) {
    adjacency[edge[0]].push(edge[1]);
  }
}

const extractStroke = (paths, endpoint_map, bridge_adjacency, log,
                       extracted_indices, start, attempt_one) => {
  const result = [];
  const visited = {};
  let current = start;

  // A list of line segments that were added to the path but that were not
  // part of the original stroke data. None of these should intersect.
  const line_segments = [];
  let self_intersecting = false;

  const advance = (index) =>
      [index[0], (index[1] + 1) % paths[index[0]].length];

  const angle = (index1, index2) => {
    const diff = Point.subtract(endpoint_map[Point.key(index2)].point,
                                endpoint_map[Point.key(index1)].point);
    assert(diff[0] !== 0 || diff[1] !== 0);
    const angle = Math.atan2(diff[1], diff[0]);
    return Angle.subtract(angle,  endpoint.angles[0]);
  }

  const getIntersection = (segment1, segment2) => {
    const diff1 = Point.subtract(segment1[1], segment1[0]);
    const diff2 = Point.subtract(segment2[1], segment2[0]);
    const cross = diff1[0]*diff2[1] - diff1[1]*diff2[0];
    if (cross === 0) {
      return undefined;
    }
    const v = Point.subtract(segment1[0], segment2[0]);
    const s = (diff1[0]*v[1] - diff1[1]*v[0])/cross;
    const t = (diff2[0]*v[1] - diff2[1]*v[0])/cross;
    if (0 < s && s < 1 && 0 < t && t < 1) {
      return [segment1[0][0] + t*diff1[0], segment1[0][1] + t*diff1[1]];
    }
    return undefined;
  }

  const indexToPoint = (index) => endpoint_map[Point.key(index)].point;

  const pushLineSegments = (points) => {
    const old_lines = line_segments.length;
    for (let i = 0; i < points.length - 1; i++) {
      line_segments.push([points[i], points[i + 1]]);
      result.push({
        start: Point.clone(points[i]),
        end: Point.clone(points[i + 1]),
        control: undefined,
      });
    }
    // Log an error if this stroke is self-intersecting.
    if (!self_intersecting) {
      for (let i = 0; i < old_lines; i++) {
        for (let j = old_lines; j < line_segments.length; j++) {
          if (getIntersection(line_segments[i], line_segments[j])) {
            self_intersecting = true;
            return;
          }
        }
      }
    }
  }

  // Here there be dragons!
  // TODO(skishore): Document the point of the geometry in this function.
  const selectBridge = (endpoint, options) => {
    if (options.length === 1 && extracted_indices[Point.key(options[0])]) {
      // Handle star-shaped strokes where one stroke ends at the intersection
      // of the bridges used by two other strokes.
      const indices1 = [endpoint.index, options[0]];
      const segment1 = indices1.map(indexToPoint);
      for (let key in bridge_adjacency) {
        if (Point.equal(endpoint_map[key].index, indices1[0])) {
          continue;
        }
        for (let i = 0; i < bridge_adjacency[key].length; i++) {
          if (Point.equal(bridge_adjacency[key][i], segment1[0])) {
            continue;
          }
          // Compute the other bridge segment and check if it intersects.
          const indices2 = [endpoint_map[key].index, bridge_adjacency[key][i]];
          const segment2 = indices2.map(indexToPoint);
          if (Point.equal(indices2[0], indices1[1]) &&
              !extracted_indices[Point.key(indices2[1])]) {
            pushLineSegments([segment1[0], segment1[1], segment2[1]]);
            return indices2[1];
          } else if (Point.equal(indices2[1], indices1[1]) &&
                     !extracted_indices[Point.key(indices2[0])]) {
            pushLineSegments([segment1[0], segment1[1], segment2[0]]);
            return indices2[0];
          }
          const intersection = getIntersection(segment1, segment2);
          if (intersection !== undefined) {
            const angle1 = angle(indices1[0], indices1[1]);
            const angle2 = angle(indices2[0], indices2[1]);
            if (Angle.subtract(angle2, angle1) < 0) {
              indices2.reverse();
              segment2.reverse();
            }
            pushLineSegments([segment1[0], intersection, segment2[1]]);
            return indices2[1];
          }
        }
      }
    } else {
      // Handle segments where the correct path is to follow a dead-end bridge,
      // even if there is another bridge that is more aligned with the stroke.
      for (let i = 0; i < options.length; i++) {
        const key = Point.key(options[i]);
        if (!extracted_indices[key]) {
          return options[i];
        }
      }
    }
    return options[0];
  }

  while (true) {
    // Add the current path segment to the path.
    result.push(paths[current[0]][current[1]]);
    visited[Point.key(current)] = true;
    current = advance(current);
    // If there are bridges at the start of the next path segment, follow the
    // one that makes the largest angle with the current path. The ordering
    // criterion enforce that we try to cross aligned bridges.
    const key = Point.key(current);
    if (bridge_adjacency.hasOwnProperty(key)) {
      var endpoint = endpoint_map[key];
      const options = bridge_adjacency[key].sort(
          (a, b) => angle(endpoint.index, a) - angle(endpoint.index, b));
      // HACK(skishore): The call to selectBridge may update the result.
      // When a stroke is formed by computing a bridge intersection, then the
      // two bridge fragments are added in selectBridge.
      const result_length = result.length;
      const next = (attempt_one ? options[0] : selectBridge(endpoint, options));
      if (result.length === result_length) {
        pushLineSegments([endpoint.point, endpoint_map[Point.key(next)].point]);
      }
      current = next;
    }
    // Check if we have either closed the loop or hit an extracted segment.
    const new_key = Point.key(current);
    if (Point.equal(current, start)) {
      if (self_intersecting) {
        log.push({cls: 'error',
                  message: 'Extracted a self-intersecting stroke.'});
      }
      let num_segments_on_path = 0;
      for (let index in visited) {
        extracted_indices[index] = true;
        num_segments_on_path += 1;
      }
      // Single-segment strokes may be due to graphical artifacts in the font.
      // We drop them to remove these artifacts.
      if (num_segments_on_path === 1) {
        log.push({cls: 'success', message: 'Dropping single-segment stroke.'});
        return undefined;
      }
      return result;
    } else if (extracted_indices[new_key] || visited[new_key]) {
      return undefined;
    }
  }
}

const extractStrokes = (paths, endpoints, bridges, log) => {
  // Build up the necessary hash tables and adjacency lists needed to run the
  // stroke extraction loop.
  const endpoint_map = {};
  const endpoint_position_map = {};
  for (let endpoint of endpoints) {
    endpoint_map[Point.key(endpoint.index)] = endpoint;
    endpoint_position_map[Point.key(endpoint.point)] = endpoint;
  }
  bridges.map(checkBridge);
  const bridge_adjacency = {};
  for (let bridge of bridges) {
    const keys = bridge.map(Point.key);
    assert(endpoint_position_map.hasOwnProperty(keys[0]));
    assert(endpoint_position_map.hasOwnProperty(keys[1]));
    const xs = keys.map((x) => endpoint_position_map[x].index);
    addEdgeToAdjacency([Point.key(xs[0]), xs[1]], bridge_adjacency);
    addEdgeToAdjacency([Point.key(xs[1]), xs[0]], bridge_adjacency);
  }
  // Actually extract strokes. Any given path segment index should appear on
  // exactly one stroke; if it is not on a stroke, we log a warning.
  const extracted_indices = {};
  const strokes = [];
  for (let attempt = 0; attempt < 3; attempt++) {
    let missed = false;
    for (var i = 0; i < paths.length; i++) {
      for (var j = 0; j < paths[i].length; j++) {
        const index = [i, j];
        if (extracted_indices[Point.key(index)]) {
          continue;
        }
        const attempt_one = attempt === 0;
        const stroke = extractStroke(paths, endpoint_map, bridge_adjacency, log,
                                     extracted_indices, index, attempt_one);
        if (stroke === undefined) {
          missed = true;
          continue;
        }
        strokes.push(stroke);
      }
    }
    if (!missed) {
      return strokes;
    }
  }
  log.push({cls: 'error',
            message: 'Stroke extraction missed some path segments.'});
  return strokes;
}

// Exports go below this fold.

const stroke_extractor = {};

stroke_extractor.getBridges = (path, classifier) => {
  const paths = svg.convertSVGPathToPaths(path);
  const endpoints = [];
  for (let i = 0; i < paths.length; i++) {
    for (let j = 0; j < paths[i].length; j++) {
      endpoints.push(new Endpoint(paths, [i, j]));
    }
  }
  classifier = classifier || stroke_extractor.combinedClassifier;
  const bridges = getBridges(endpoints, classifier);
  return {endpoints: endpoints, bridges: bridges};
}

stroke_extractor.getStrokes = (path, bridges) => {
  const paths = svg.convertSVGPathToPaths(path);
  const endpoints = [];
  for (let i = 0; i < paths.length; i++) {
    for (let j = 0; j < paths[i].length; j++) {
      endpoints.push(new Endpoint(paths, [i, j]));
    }
  }
  const log = [];
  const stroke_paths = extractStrokes(paths, endpoints, bridges, log);
  const strokes = stroke_paths.map((x) => svg.convertPathsToSVGPath([x]));
  return {log: log, strokes: strokes};
}

stroke_extractor.handTunedClassifier = handTunedClassifier;

export {stroke_extractor};
